pkg/rtc/subscriptionmanager_test.go

```diff
@@ -0,0 +1,386 @@
+ * limitations under the License.
+ */
+
+package rtc
+
+import (
+	"sync"
+	"sync/atomic"
+	"testing"
+	"time"
+
+	"github.com/stretchr/testify/require"
+
+	"github.com/livekit/livekit-server/pkg/rtc/types"
+	"github.com/livekit/livekit-server/pkg/rtc/types/typesfakes"
+	"github.com/livekit/livekit-server/pkg/telemetry/telemetryfakes"
+	"github.com/livekit/protocol/livekit"
+	"github.com/livekit/protocol/logger"
+)
+
+func init() {
+	reconcileInterval = 50 * time.Millisecond
+	notFoundTimeout = 200 * time.Millisecond
+	subscriptionTimeout = 200 * time.Millisecond
+}
+
+const (
+	subSettleTimeout = 300 * time.Millisecond
+	subCheckInterval = 10 * time.Millisecond
+)
+
+func TestSubscribe(t *testing.T) {
+	t.Run("happy path subscribe", func(t *testing.T) {
+		sm := newTestSubscriptionManager(t)
+		defer sm.Close(false)
+		resolver := newTestResolver(true, nil)
+		sm.params.TrackResolver = resolver.Resolve
+		subCount := atomic.Int32{}
+		failed := atomic.Bool{}
+		sm.params.OnTrackSubscribed = func(subTrack types.SubscribedTrack) {
+			subCount.Add(1)
+		}
+		sm.params.OnSubcriptionError = func(trackID livekit.TrackID) {
+			failed.Store(true)
+		}
+		numParticipantSubscribed := atomic.Int32{}
+		numParticipantUnsubscribed := atomic.Int32{}
+		sm.OnSubscribeStatusChanged(func(pubID livekit.ParticipantID, subscribed bool) {
+			if subscribed {
+				numParticipantSubscribed.Add(1)
+			} else {
+				numParticipantUnsubscribed.Add(1)
+			}
+		})
+
+		sm.SubscribeToTrack("track", "pub", "pubID")
+		s := sm.subscriptions["track"]
+		require.True(t, s.isDesired())
+		require.Nil(t, s.getSubscribedTrack())
+		require.Eventually(t, func() bool {
+			return subCount.Load() == 1
+		}, subSettleTimeout, subCheckInterval, "track was not subscribed")
+
+		require.Len(t, sm.GetSubscribedTracks(), 1)
+		require.Len(t, sm.GetSubscribedParticipants(), 1)
+		require.Equal(t, "pubID", string(sm.GetSubscribedParticipants()[0]))
+
+		// ensure telemetry events are sent
+		tm := sm.params.Telemetry.(*telemetryfakes.FakeTelemetryService)
+		require.Equal(t, 1, tm.TrackSubscribeRequestedCallCount())
+
+		// ensure bound
+		setTestSubscribedTrackBound(t, s.getSubscribedTrack())
+
+		require.Eventually(t, func() bool {
+			return !s.needsBind()
+		}, subSettleTimeout, subCheckInterval, "track was not bound")
+
+		// telemetry event should have been sent
+		require.Equal(t, 1, tm.TrackSubscribedCallCount())
+
+		time.Sleep(notFoundTimeout)
+		require.False(t, failed.Load())
+
+		// ensure its resilience after being closed
+		setTestSubscribedTrackClosed(t, s.getSubscribedTrack(), false)
+		require.True(t, s.needsSubscribe())
+
+		require.Eventually(t, func() bool {
+			return s.isDesired() && !s.needsSubscribe()
+		}, subSettleTimeout, subCheckInterval, "track was not resubscribed")
+
+		// was subscribed twice, unsubscribed once (due to close)
+		require.Equal(t, int32(2), numParticipantSubscribed.Load())
```

Comment:
this line will currently trigger the subscription to close

```
setTestSubscribedTrackClosed(t, s.getSubscribedTrack(), false)
```

for the participants subscribed to, we could either go with desired vs actual state. I went with the latter. So in this test, the subscribed track closure caused the subscription to be unsubscribed. That made it so that we do not perceive the participant to be actually subscribed to the publisher any longer.

however, reconciler would repair that condition and recreated the subscription.

Reply:
I kinda guessed that's what is happening 😄 . But, how could a subscription succeed again through the reconciler path if the published track is closed. Shouldn't the resolver return an empty track? That's what confused me, i. e. how did the reconciler make a successful subscription on a closed track?

Reply:
The published track wasn't closed, it was only the subscribedTrack. It's just a simulated test case to ensure reconciler is resilient to recreate the subscription.
In practice, the mediaTrack could have been closed, and this path will keep waiting until resolver could resolve the track again.

Reply:
oh got it. Thank you